core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java

```diff
@@ -149,22 +159,24 @@ public UnderFileSystem addMount(long mountId, String ufsUri, Map<String, String>
       throws IOException {
     Preconditions.checkArgument(mountId != IdUtils.INVALID_MOUNT_ID, "mountId");
     Preconditions.checkArgument(ufsUri != null, "uri");
-    UnderFileSystem ufs = getOrAdd(ufsUri, ufsConf);
-    mMountIdToUnderFileSystemMap.put(mountId, ufs);
-    return ufs;
+    Pair<UnderFileSystem, AtomicLong> ufsCounter = getOrAdd(ufsUri, ufsConf);
+    mMountIdToUnderFileSystemMap.put(mountId, ufsCounter);
+    return ufsCounter.getFirst();
   }
 
   @Override
   public void removeMount(long mountId) {
     Preconditions.checkArgument(mountId != IdUtils.INVALID_MOUNT_ID, "mountId");
-    // TODO(binfan): check the refcount of this ufs in mUnderFileSystemMap and remove it if this is
-    // no more used. Currently, it is possibly used by out mount too.
-    mMountIdToUnderFileSystemMap.remove(mountId);
+    Pair<UnderFileSystem, AtomicLong> ufsCounter = mMountIdToUnderFileSystemMap.get(mountId);
+    // Remove ufs from mMountIdToUnderFileSystemMap if its counter comes to zero.
+    if (ufsCounter.getSecond().decrementAndGet() == 0L) {
+      mMountIdToUnderFileSystemMap.remove(mountId);
+    }
   }
 
   @Override
   public UnderFileSystem get(long mountId) {
-    return mMountIdToUnderFileSystemMap.get(mountId);
+    return mMountIdToUnderFileSystemMap.get(mountId).getFirst();
```

Comment:
maybe I am not clear here. Let me rephrase:
in case `mountId` doesn't exist at all in `mMountIdToUnderFileSystemMap`, original implementation will return null as this map contains no mapping for the key, and this method will return null. While the current way changes this behavior: if `mountId` doesn't exist, `mMountIdToUnderFileSystemMap.get(mountId)` will return null still, but `null.getFirst()` will get a `NullPointerException`. So the way to keep the same behavior can be something like this:

```java
Pair<UnderFileSystem, AtomicLong> counter = mMountIdToUnderFileSystemMap.get(mountId);
if (counter == null) {
  return null;
}
return counter.getFirst();
```